### ISSUE TITLE
RemovePodsViolatingTopologySpreadConstraint restarting pods on a loop

### ISSUE THREAD

**Issue by dk-d-n** [kubernetes-sigs/descheduler#839]
**What version of descheduler are you using?**

descheduler version: 0.22.1

**Does this issue reproduce with the latest release?**
Yes

**Which descheduler CLI options are you using?**
            - --policy-config-file
            - /policy-dir/policy.yaml
            - --v
            - "5"

**Please provide a copy of your descheduler policy config file**
`apiVersion: "descheduler/v1alpha1"
kind: "DeschedulerPolicy"
strategies:
  LowNodeUtilization:
    enabled: false
  RemoveDuplicates:
    enabled: false
  RemovePodsViolatingInterPodAntiAffinity:
    enabled: false
  RemovePodsViolatingNodeAffinity:
    enabled: false
  RemovePodsViolatingNodeTaints:
    enabled: false
  RemovePodsViolatingTopologySpreadConstraint:
    enabled: true`

**What k8s version are you using (`kubectl version`)?**
1.21.8

**What did you do?**

I have a cluster that contains 3 groups of nodes:
1) Control Plane
2) Infrastructure Services
3) Application (2 nodes)

All of them have taints.  On deployment, my application helm chart will deploy 4 pods to my application node group correctly using nodeAffinity on the application node labels with a maxSkew of 1, `maxSkew=1, whenUnsatisfiable=ScheduleAnyways`. 2 pods on each application node.  Descheduler, which runs every hour, will calculate some skew and evict 2 of the 4 application pods, causing it to spin up once again on the same nodes.  If I use the nodeSelector param and target only my application nodes using its label, it will detect my topology as balanced and not evict any of my pods.  I've tried using nodeFit as well.

**What did you expect to see?**
No change should happen.

**What did you see instead?**
Descheduler, which runs every hour, will calculate some skew and evict 2 of the 4 application pods, causing it to spin up once again on the same nodes.

Are my other nodes with taints being considered as part of the algorithm?  Can't tell if this is expected behavior.

**Comment by dk-d-n** [comment]
For each node in my cluster that does not have the application taint the following is outputted:
```
I0606 20:56:18.523151 1 node.go:148] "Pod does not fit on node" pod="<podName>" node="<nodeName>"
I0606 20:56:18.523157 1 node.go:150] "pod node selector does not match the node label"
I0606 20:56:18.523162 1 node.go:150] "pod does not tolerate taints on the node"
```

**Comment by a7i** [comment]
>Are my other nodes with taints being considered as part of the algorithm? Can't tell if this is expected behavior.

No, the nodeFit algorithm should exclude the nodes which the Pod does not tolerate.

Can you please share the following:

- Taints definition of the nodes
- Your scheduler policy with nodeFit enabled

Lastly, I am surprised that it is evicting on `ScheduleAnyways` when `includeSoftConstraints` is not set to true.

**Comment by dk-d-n** [comment]
Control Plane Taints:
```
[
  {
    "effect": "NoSchedule",
    "key": "node-role.kubernetes.io/controlplane",
    "value": "true"
  },
  {
    "effect": "NoExecute",
    "key": "node-role.kubernetes.io/etcd",
    "value": "true"
  }
]
```

Infra Service Taints:
```
  {
    "effect": "NoSchedule",
    "key": "infra_service"
  }
```

Our application nodes do not have taints but our app base helm chart are configured with constraints and affinity that look like the following:
```
      topologySpreadConstraints:
      - labelSelector:
          matchLabels:
            app.kubernetes.io/instance: service
            app.kubernetes.io/name: service
        maxSkew: 1
        topologyKey: kubernetes.io/hostname
        whenUnsatisfiable: DoNotSchedule
....
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: app_service
                operator: Exists
```

Node Labels:
```
## Control Plane
beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,cattle.io/creator=norman,kubernetes.io/arch=amd64,kubernetes.io/hostname=cp-1,kubernetes.io/os=linux,node-role.kubernetes.io/controlplane=true,node-role.kubernetes.io/etcd=true

## Application Node 
beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,cattle.io/creator=norman,app_service=true,kubernetes.io/arch=amd64,kubernetes.io/hostname=app-1,kubernetes.io/os=linux,node-role.kubernetes.io/worker=true

## Infra Node
beta.kubernetes.io/arch=amd64,beta.kubernetes.io/os=linux,cattle.io/creator=norman,infra_service=true,kubernetes.io/arch=amd64,kubernetes.io/hostname=infra-1,kubernetes.io/os=linux,node-role.kubernetes.io/worker=true
```

Descheduler Policy:
```
apiVersion: "descheduler/v1alpha1"
kind: "DeschedulerPolicy"
strategies:
  LowNodeUtilization:
    enabled: false
  RemoveDuplicates:
    enabled: false
  RemovePodsViolatingInterPodAntiAffinity:
    enabled: false
  RemovePodsViolatingNodeAffinity:
    enabled: false
  RemovePodsViolatingNodeTaints:
    enabled: false
  RemovePodsViolatingTopologySpreadConstraint:
    enabled: true
    params:
      includeSoftConstraints: false
      nodeFit: true
```

Can confirm that nodeFit option still evicts the pods every run.

**Comment by a7i** [comment]
Thanks for providing all the info, it is very helpful!

/assign

**Comment by dk-d-n** [comment]
Just wanted to point out that in my initial post i had a soft constraint `whenUnsatisfiable=ScheduleAnyways`.  This is actually `whenUnsatisfiable: DoNotSchedule` as posted correctly in the latter post.

**Comment by a7i** [comment]
Just wanted to report that I am able to reproduce this.

I added a unit test below which is failing (it's eviction 2 Pods but the expected count is zero): https://github.com/a7i/descheduler/commit/c158d1c06c0061ff6f2cf7497c695ed29ef4b2c0

I can also confirm that PR #836 does resolve the issue but it's still under review 🤞🏼

**Comment by a7i** [comment]
@dk-d-n the fix has been merged (but not yet released). Would you be able to test this scenario using the latest on the main branch?

I have added a test to cover the scenario but wanted your confirmation before closing this Issue.

**Comment by a7i** [comment]
The issue was addressed in 0.25

Please feel free to open if that is not the case @dk-d-n 

/close
